base/save.go

```diff
@@ -269,3 +252,75 @@ func GenerateAvailableName(r repo.Repo, peername, prefix string) string {
 		}
 	}
 }
+
+// DatasetNameExists determines whether the name exists in the repository
+// TODO(dustmop): Add dscache support
+func DatasetNameExists(r repo.Repo, dsName string) bool {
+	pro, err := r.Profile()
+	if err != nil {
+		log.Errorf("couldn't get profile: %s", err)
+		return false
+	}
+	peername := pro.Peername
+	lookup := &reporef.DatasetRef{Name: dsName, Peername: peername}
+	err = repo.CanonicalizeDatasetRef(r, lookup)
+	if err == repo.ErrNotFound {
+		return false
+	}
+	return true
+}
+
+// DatasetNameAndStableIdentifers determines the final name for the dataset, by inferring one if
+// necessary, and returns stable identifiers for the full dataset history, and for the most
+// recent version.
+func DatasetNameAndStableIdentifers(ctx context.Context, r repo.Repo, peername string, dsName string, ds *dataset.Dataset, newName bool) (string, string, string, error) {
```

Comment:
I think this should return `(dsref.Ref, error)` and be destructured up in lib.

Reply:
That's a very good idea. This requires added the `InitID` field to dsref, which https://github.com/qri-io/qri/pull/1308 also does, so I'm going to copy it over to this PR.